Funcky.DiscriminatedUnion.SourceGeneration/Emitter.cs

```diff
@@ -18,25 +17,81 @@ public static string EmitDiscriminatedUnion(DiscriminatedUnion discriminatedUnio
-            writer.OpenScope();
-
-            WriteGeneratedMethod(writer, $"{discriminatedUnion.MethodVisibility} abstract {FormatMatchMethodDeclaration(discriminatedUnion.MatchResultTypeName, discriminatedUnion.Variants)};");
-            writer.WriteLine();
-            WriteGeneratedMethod(writer, $"{discriminatedUnion.MethodVisibility} abstract {FormatSwitchMethodDeclaration(discriminatedUnion.Variants)};");
-
-            foreach (var variant in discriminatedUnion.Variants)
+            if (discriminatedUnion.GeneratePartitionExtension)
             {
-                WriteVariant(writer, discriminatedUnion, variant);
+                writer.WriteLine();
+                WritePartitionExtension(discriminatedUnion, writer);
             }
         }
 
         return stringBuilder.ToString();
     }
 
+    private static void WriteUnionType(DiscriminatedUnion discriminatedUnion, IndentedTextWriter writer)
+    {
+        using var scope = writer.AutoCloseScopes();
+
+        WriteParentTypes(writer, discriminatedUnion.ParentTypes);
+
+        WriteJsonDerivedTypeAttributes(writer, discriminatedUnion);
+        writer.WriteLine(FormatPartialTypeDeclaration(discriminatedUnion.Type));
+        writer.OpenScope();
+
+        WriteGeneratedMethod(writer, $"{discriminatedUnion.MethodVisibility} abstract {FormatMatchMethodDeclaration(discriminatedUnion.MatchResultTypeName, discriminatedUnion.Variants)};");
+        writer.WriteLine();
+        WriteGeneratedMethod(writer, $"{discriminatedUnion.MethodVisibility} abstract {FormatSwitchMethodDeclaration(discriminatedUnion.Variants)};");
+
+        foreach (var variant in discriminatedUnion.Variants)
+        {
+            WriteVariant(writer, discriminatedUnion, variant);
+        }
+    }
+
+    private static void WritePartitionExtension(DiscriminatedUnion discriminatedUnion, IndentedTextWriter writer)
+    {
+        using var scope = writer.AutoCloseScopes();
+
+        writer.WriteLine($"{discriminatedUnion.MethodVisibility} static class {discriminatedUnion.Type.Identifier}EnumerableExtensions");
+        writer.OpenScope();
+
+        writer.Write("public record struct Partitions(");
+        var partitionVariants = discriminatedUnion
+            .Variants
+            .Select(v => $"System.Collections.Generic.IReadOnlyList<{discriminatedUnion.Type.Identifier}.{v.LocalTypeName}> {v.ParameterName}");
+        writer.Write(string.Join(", ", partitionVariants));
+        writer.WriteLine(");");
+
+        writer.WriteLine();
+
+        writer.WriteLine($"public static Partitions Partition(this System.Collections.Generic.IEnumerable<{discriminatedUnion.Type.Identifier}> source)");
+        writer.OpenScope();
+
+        foreach (var variant in discriminatedUnion.Variants)
+        {
+            writer.WriteLine($"var {variant.ParameterName}Items = System.Collections.Immutable.ImmutableList.CreateBuilder<{discriminatedUnion.Type.Identifier}.{variant.LocalTypeName}>();");
```

Comment:
Is the immutable collection necessary here? It adds an additional dependency on older target frameworks for little gain. You could just use a regular ol' list and use `AsReadOnly` to enforce the readonly-ness.

Reply:
Good idea, done